base/acme/src/main/java/org/dogtagpki/acme/validator/HTTP01Validator.java

```diff
@@ -47,49 +45,34 @@ public void validateChallenge(
         ACMEIdentifier identifier = authorization.getIdentifier();
         String hostname = identifier.getValue();
         String validationPath = "/.well-known/acme-challenge/" + token;
-        URI validationURL = new URI("http", hostname, validationPath, null);
-
-        // TODO: move retry to ACMEChallengeProcessor.processChallenge()
-        // TODO: make it configurable
-
-        int maxCount = 5;
-        int interval = 5;
-
-        int count = 1;
-        String response = null;
-        Exception exception = null;
-
-        while (true) {
-            try {
-                response = getResponse(validationURL);
-                break;
-
-            } catch (Exception e) {
-
-                // TODO: catch more specific HTTP exception
 
-                logger.error(e.getMessage());
-                if (count >= maxCount) {
-                    exception = e;
-                    break;
-                }
-
-                Thread.sleep(interval * 1000);
-                count++;
-            }
+        URI validationURL;
+        try {
+            validationURL = new URI("http", hostname, validationPath, null);
+        } catch (Throwable e) {
```

Comment:
I disagree.  `hostname` comes from the identifier object, via the authorization object.  `validationPath` is derived from `token` which is retrieved from the challenge object.  All of these objects are looked up server-side.  If we got this far already, the request was fine, but something is messed up on the server side.  `serverInternal` is the correct error.  I will add a comment to the code to explain this.

Reply:
The `URI` constructor only declares `URISyntaxException`, so that's the only exception we can catch from the above code, unless we want to catch something more generic such as `Exception` or `RuntimeException` (but not `Throwable` please).

The `URISyntaxException` can happen if the `token` or the `hostname` is malformed. The `token` is unlikely to be malformed unless there's a bug in the code (which can be reduced through testing), but the `hostname` is provided by the user and has not been validated by the server at this point, so there's always a chance of error here. It would be useful for the users to tell them that there's a problem with the `hostname` they provided. The `serverInternal` could mislead the users unless they actually inspect the error message.

Reply:
@edewata if hostname has not been validated yet, would should validate it sooner, before the order is created.

Reply:
@edewata I propose to add a validation method to `ACMEIdentifier` to check the syntax of the identifier.  It should be invoked in `ACMENewOrderService` and if the syntax is not valid, the order should not be created; `malformed` error should be returned at that point.

Reply:
I thought the ACME challenge is supposed to validate not only the hostname ownership but also the validity of the hostname itself (because otherwise the validation would fail), so it's not necessary to do a separate hostname validation.

Suppose we implement the hostname validation, what validation rules would we use? Is this defined in RFC 8555? Also, I don't think it should be implemented in `ACMEIdentifier` since it's a POJO class. I suppose some people might want to use different hostname rules for their specific environment so it probably should be customizable somehow in `ACMEEngine` class.

Reply:
@edewata the syntactic validity of a hostname is well defined!  RFC 8555 (ACME) says do it according to RFC 5280 (X.509) which in turn says use the preferred name syntax from RFC 1034 (DNS).  ACME additionally allows a leftmost wildcard label.

If an identifier doesn't even look like a DNS name, then we should summarily reject the order.  This behaviour is implied by way of a specific example in https://tools.ietf.org/html/rfc8555#section-6.7.1.  A `newOrder` request with malformed identifier fails with status 4xx:

```
   ACME clients may choose to use the "identifier" field of a subproblem
   as a hint that an operation would succeed if that identifier were
   omitted.  For instance, if an order contains ten DNS identifiers, and
   the newOrder request returns a problem document with two subproblems
   (referencing two of those identifiers), the ACME client may choose to
   submit another order containing only the eight identifiers not listed
   in the problem document.

HTTP/1.1 403 Forbidden
Content-Type: application/problem+json
Link: <https://example.com/acme/directory>;rel="index"

{
    "type": "urn:ietf:params:acme:error:malformed",
    "detail": "Some of the identifiers requested were rejected",
    "subproblems": [
        {
            "type": "urn:ietf:params:acme:error:malformed",
            "detail": "Invalid underscore in DNS name \"_example.org\"",
            "identifier": {
                "type": "dns",
                "value": "_example.org"
            }
        },
        {
            "type": "urn:ietf:params:acme:error:rejectedIdentifier",
            "detail": "This CA will not issue for \"example.net\"",
            "identifier": {
                "type": "dns",
                "value": "example.net"
            }
        }
    ]
}
```

I really don't think DNS name validation needs to be customisable.  Certainly we can reject it if it's not syntactically valid.  The `ACMEPolicy` framework provides the additional flexibility needed to further constrain issuance according to organisational requirements.

Anyhow, I implemented the syntax checking; see the latest commits in this PR.